sympy/polys/matrices/linsolve.py

```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
+    sdm_aug = SDM(enumerate(eqsdict), (neqs, nsyms + 1), K)
     return sdm_aug
 
 
-def _expand_eqs_deprecated(eqs):
-    """Use expand to cancel nonlinear terms.
+def _linear_eq_to_dict(eqs, syms):
+    """Convert a system Expr/Eq equations into dict form, returning
+    the coefficient dictionaries and a list of syms-independent terms
+    from each expression in ``eqs```.
 
-    This approach matches previous behaviour of linsolve but should be
-    deprecated.
+    Examples
+    ========
+
+    >>> from sympy.polys.matrices.linsolve import _linear_eq_to_dict
+    >>> from sympy.abc import x
+    >>> _linear_eq_to_dict([2*x + 3], {x})
+    ([{x: 2}], [3])
     """
-    def expand_eq(eq):
-        if eq.is_Equality:
-            eq = eq.lhs - eq.rhs
-        return eq.expand()
+    coeffs = []
+    ind = []
+    symset = set(syms)
+    for i, e in enumerate(eqs):
+        if e.is_Equality:
+            (coeff, terms), (cR, tR) = [_lin_eq2dict(ai, symset)
+                for ai in e.args]
+            # there were no nonlinear errors so now
+            # cancellation is allowed
+            coeff -= cR
+            for k, v in tR.items():
+                if k in terms:
+                    terms[k] -= v
+                else:
+                    terms[k] = v
+            # don't store coefficients of 0, however
+            terms = {k: v for k, v in terms.items() if v}
+            c, d = coeff, terms
+        else:
+            c, d = _lin_eq2dict(e, symset)
+        coeffs.append(d)
+        ind.append(c)
+    return coeffs, ind
 
-    return [expand_eq(eq) for eq in eqs]
 
+def _lin_eq2dict(a, symset):
+    """return (c, d) where c is the sym-independent part of ``a`` and
+    ``d`` is an efficiently calculated dictionary mapping symbols to
+    their coefficients. A PolyNonlinearError is raised if non-linearity
+    is detected.
 
-def _linear_eq_to_dict(eqs, syms):
-    """Convert a system Expr/Eq equations into dict form"""
-    try:
-        return _linear_eq_to_dict_inner(eqs, syms)
-    except PolyNonlinearError:
-        # XXX: This should be deprecated:
-        eqs = _expand_eqs_deprecated(eqs)
-        return _linear_eq_to_dict_inner(eqs, syms)
-
-
-def _linear_eq_to_dict_inner(eqs, syms):
-    """Convert a system Expr/Eq equations into dict form"""
-    syms = set(syms)
-    eqsdict, eqs_rhs = [], []
-    for eq in eqs:
-        rhs, eqdict = _lin_eq2dict(eq, syms)
-        eqsdict.append(eqdict)
-        eqs_rhs.append(rhs)
-    return eqsdict, eqs_rhs
+    The values in the dictionary will be non-zero.
 
+    Examples
+    ========
 
-def _lin_eq2dict(a, symset):
-    """Efficiently convert a linear equation to a dict of coefficients"""
+    >>> from sympy.polys.matrices.linsolve import _lin_eq2dict
+    >>> from sympy.abc import x, y
+    >>> _lin_eq2dict(x + 2*y + 3, {x, y})
+    (3, {x: 1, y: 2})
+    """
     if a in symset:
         return S.Zero, {a: S.One}
-    elif a.is_Add:
+    if a.is_Add:
```

Comment:
Why have you changed these `elif` to `if`? It's much clearer to use `elif` to show that all cases are mutually exclusive without needing to scan through to see that every control path either returns or raises.

Reply:
Preference. If every block clearly ends with a return (or starts with an elif) coverage is complete.